### PR TITLE
[change] Remove the action_holder placeholder from the policy. 

### PR DESCRIPTION
**Pull request by ervteng**
### Proposed change(s)

In the current implementation, the Policy has a placeholder called an "action_holder" that allows you to feed in an action externally during a `sess.run` call. This is used to e.g. compute the one-hot action in discrete, to feed in the continuous action to the Q function in SAC, and to compute the entropy.

However, TF allows you to feed in a value for _any_ tensor, not just placeholders. Thus, in this PR, we just feed the value into the output tensor, which removes the need for action_holders. This makes implementing the output as a separate Distribution object much easier. 

### Types of change(s)

- [ ] Bug fix
- [ ] New feature
- [x] Code refactor
- [ ] Breaking change
- [ ] Documentation update
- [ ] Other (please describe)

### Checklist
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added updated the changelog (if applicable)
- [ ] I have added necessary documentation (if applicable)
- [ ] I have updated the migration guide (if applicable)

### CODE REVIEW

File: ml-agents/mlagents/trainers/common/nn_policy.py [hunk 1]
```diff
@@ -276,10 +276,6 @@ def _create_cc_actor(
             (tf.identity(self.all_log_probs)), axis=1, keepdims=True
         )
 
-        self.action_holder = tf.placeholder(
-            shape=[None, self.act_size[0]], dtype=tf.float32, name="action_holder"
-        )
-
     def _create_dc_actor(self, encoded: tf.Tensor) -> None:
         """
         Creates Discrete control actor-critic model.
```
Comment (andrewcoh) [inline ml-agents/mlagents/trainers/common/nn_policy.py]: param docsrings don't match

Comment (andrewcoh) [inline ml-agents/mlagents/trainers/common/nn_policy.py]: param docstrings don't match

File: ml-agents/mlagents/trainers/components/reward_signals/gail/signal.py [hunk 2]
```diff
@@ -82,7 +82,7 @@ def evaluate_batch(self, mini_batch: Dict[str, np.array]) -> RewardSignalResult:
         if self.policy.use_continuous_act:
             feed_dict[self.policy.selected_actions] = mini_batch["actions"]
         else:
-            feed_dict[self.policy.action_holder] = mini_batch["actions"]
+            feed_dict[self.policy.output] = mini_batch["actions"]
```
Comment (andrewcoh) [inline ml-agents/mlagents/trainers/components/reward_signals/gail/signal.py]: feels like we are overloading `self.policy.output` to act as both a placeholder self.policy.action_holder and also the tensor that represents the tensorflow function that composes the policy (since we backprop through it when reparameterizing).

File: ml-agents/mlagents/trainers/sac/network.py [hunk 3]
```diff
@@ -121,10 +121,14 @@ def _create_cc_critic(self, hidden_value, scope, create_qs=True):
             self.h_size,
             self.join_scopes(scope, "value"),
         )
-
+        self.external_action_in = tf.placeholder(
```
Comment (andrewcoh) [inline ml-agents/mlagents/trainers/sac/network.py]: Does this change just shift the burden of who needs to have the placeholder?

File: ml-agents/mlagents/trainers/sac/optimizer.py [hunk 4]
```diff
@@ -171,15 +171,14 @@ def _create_inputs_and_outputs(self) -> None:
...
         self.next_visual_in = self.target_network.visual_in
-        self.action_holder = self.policy.action_holder
         self.sequence_length_ph = self.policy.sequence_length_ph
         self.next_sequence_length_ph = self.target_network.sequence_length_ph
         if not self.policy.use_continuous_act:
             self.action_masks = self.policy_network.action_masks
         else:
             self.output_pre = self.policy_network.output_pre
 
-        # Don't use value estimate during inference. TODO: Check why PPO uses value_estimate in inference.
```
Comment (vincentpierre) [inline ml-agents/mlagents/trainers/sac/optimizer.py]: Is this TODO still todo ?

File: ml-agents/mlagents/trainers/tests/test_meta_curriculum.py [hunk 5]
```diff
@@ -102,7 +102,7 @@ def test_get_config():
         hidden_units: 128
         lambd: 0.95
         learning_rate: 5.0e-3
-        max_steps: 300
+        max_steps: 100
```
Comment (vincentpierre) [inline ml-agents/mlagents/trainers/tests/test_meta_curriculum.py]: Why does the max_step need to change ?

File: ml-agents/mlagents/trainers/tests/test_simple_rl.py [hunk 6]
```diff
@@ -263,9 +263,12 @@ def _check_environment_trains(
         # Begin training
         tc.start_learning(env_manager)
         print(tc._get_measure_vals())
-        for mean_reward in tc._get_measure_vals().values():
```
Comment (vincentpierre) [inline ml-agents/mlagents/trainers/tests/test_simple_rl.py]: Seems this change should be part of another PR

File: ml-agents/mlagents/trainers/tests/test_meta_curriculum.py (hunk 5 above)
Comment (ervteng) [reply ml-agents/mlagents/trainers/tests/test_meta_curriculum.py]: Reduce the amount of time taken for this test. I can break this out into a separate PR but it was needed to get the tests to pass for this one. 

Basically depending on the seed there was a chance this test would not complete an episode. So some PRs it passed and others it didn't. I'll branch this out into a separate PR.

File: ml-agents/mlagents/trainers/sac/network.py (hunk 3 above)
Comment (ervteng) [reply ml-agents/mlagents/trainers/sac/network.py]: Yes, for SAC we can get away with moving the placeholder since it wasn't connected anywhere further down the graph.

File: ml-agents/mlagents/trainers/sac/optimizer.py (hunk 4 above)
Comment (ervteng) [reply ml-agents/mlagents/trainers/sac/optimizer.py]: Nope, I removed the value estimate during inference in the previous PR.

File: ml-agents/mlagents/trainers/tests/test_simple_rl.py (hunk 6 above)
Comment (ervteng) [reply ml-agents/mlagents/trainers/tests/test_simple_rl.py]: Moved to separate PR #3511
